src/split_at.rs

```diff
@@ -233,59 +213,571 @@ unsafe impl<T> SplitAt for [T] {
-
-    // SAFETY:
-    // 0. `left` conforms to the aliasing invariant of `I::Aliasing, by Lemma 0.
-    // 1. `left` conforms to the alignment invariant of `I::Alignment, because
-    //    the referents of `left` and `Self` have the same address and type
-    //    (and, thus, alignment requirement).
-    // 2. `left` conforms to the validity invariant of `I::Validity`, neither
-    //    the type nor bytes of `left`'s referent have been changed.
-    let left = unsafe { Ptr::from_inner(left) };
-
-    // SAFETY:
-    // 0. `right` conforms to the aliasing invariant of `I::Aliasing, by Lemma
-    //    0.
-    // 1. `right` conforms to the alignment invariant of `I::Alignment, because
-    //    if `ptr` with `I::Alignment = Aligned`, then by invariant on `T:
-    //    SplitAt`, the trailing slice of `ptr` (from which `right` is derived)
-    //    will also be well-aligned.
-    // 2. `right` conforms to the validity invariant of `I::Validity`, because
-    //    `right: [T::Elem]` is derived from the trailing slice of `ptr`, which,
-    //    by contract on `T: SplitAt::Elem`, has type `[T::Elem]`.
-    let right = unsafe { Ptr::from_inner(right) };
-
-    (left, right)
+    /// Produces the length of `self`'s left part.
+    #[inline(always)]
+    fn l_len(&self) -> MetadataOf<T> {
+        // SAFETY: By invariant on `Split`, `self.l_len` is not greater than the
+        // length of `self.source`.
+        unsafe { MetadataOf::<T>::new_unchecked(self.l_len) }
+    }
+
+    /// Produces the split parts of `self`, using [`Immutable`] to ensure that
+    /// it is sound to have concurrent references to both parts.
+    #[inline(always)]
+    fn via_immutable(self) -> (Ptr<'a, T, I>, Ptr<'a, [T::Elem], I>)
+    where
+        T: Immutable,
+        I: Invariants<Aliasing = Shared>,
+    {
+        // SAFETY: `Aliasing = Shared` and `T: Immutable`.
+        unsafe { self.via_unchecked() }
+    }
+
+    /// Produces the split parts of `self`, using [`IntoBytes`] to ensure that
+    /// it is sound to have concurrent references to both parts.
+    #[inline(always)]
+    fn via_into_bytes(self) -> (Ptr<'a, T, I>, Ptr<'a, [T::Elem], I>)
+    where
+        T: IntoBytes,
+    {
+        // SAFETY: By `T: IntoBytes`, `T` has no padding for any length.
+        // Consequently, `T` can be split into non-overlapping parts at any
+        // index.
+        unsafe { self.via_unchecked() }
+    }
+
+    /// Produces the split parts of `self`, using [`Unaligned`] to ensure that
+    /// it is sound to have concurrent references to both parts.
+    #[inline(always)]
+    fn via_unaligned(self) -> (Ptr<'a, T, I>, Ptr<'a, [T::Elem], I>)
+    where
+        T: Unaligned,
+    {
+        // SAFETY: By `T: SplitAt + Unaligned`, `T` is either `[T]`, or a
+        // `repr(C)` or `repr(transparent)` slice DST, that is well-aligned at
+        // any address and length. Consequently, `T` can be split into strictly
+        // non-overlapping parts any any index.
```

Comment:
Can you more explicitly argue that `Unaligned + repr(C)` implies that there will be no padding? IMO you have to be quite familiar with our prior discussions to understand the logical step happening here.

Reply:
https://github.com/google/zerocopy/compare/9562f5e1784e47002f2e20edbdc6cfd7cb753afe..5ae44a7b575613bc5ff777c1d84649fe55679889